Source/Diagnostics/ReducedDiags/FieldReduction.cpp

```diff
@@ -0,0 +1,98 @@
+/* Copyright 2021 Neil Zaim
+ *
+ * This file is part of WarpX.
+ *
+ * License: BSD-3-Clause-LBNL
+ */
+
+#include "FieldReduction.H"
+#include "WarpX.H"
+#include "Utils/WarpXAlgorithmSelection.H"
+
+// constructor
+FieldReduction::FieldReduction (std::string rd_name)
+: ReducedDiags{rd_name}
+{
+    using namespace amrex::literals;
+
+    // RZ coordinate is not working
+#if (defined WARPX_DIM_RZ)
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(false,
+        "FieldReduction reduced diagnostics does not work for RZ coordinate.");
+#endif
+
+    // read number of levels
+    int nLevel = 0;
+    amrex::ParmParse pp_amr("amr");
+    pp_amr.query("max_level", nLevel);
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(nLevel == 0,
+        "FieldReduction reduced diagnostics does not work with mesh refinement.");
+
+    constexpr int noutputs = 1; // A single output in the Field reduction diagnostic
+    // resize data array
+    m_data.resize(noutputs, 0.0_rt);
+
+    amrex::ParmParse pp_rd_name(rd_name);
+
+    // read reduced function with parser
+    std::string parser_string = "";
+    Store_parserString(pp_rd_name,"reduced_function(x,y,z,Ex,Ey,Ez,Bx,By,Bz)",
+                       parser_string);
+    // Remove newlines and spaces in the parser string. This is because this string is written in
+    // the header of the reduced diag output file so this should prevent weird formatting from
+    // occuring.
+    parser_string.erase(std::remove(parser_string.begin(), parser_string.end(), '\n'),
+                        parser_string.end());
+    parser_string.erase(std::remove(parser_string.begin(), parser_string.end(), ' '),
+                        parser_string.end());
+    m_parser = std::make_unique<ParserWrapper<m_nvars>>(
+        makeParser(parser_string,{"x","y","z","Ex","Ey","Ez","Bx","By","Bz"}));
+
+    // read reduction type
+    std::string reduction_type_string;
+    pp_rd_name.get("reduction_type", reduction_type_string);
+    m_reduction_type = GetAlgorithmInteger (pp_rd_name, "reduction_type");
+
+    if (amrex::ParallelDescriptor::IOProcessor())
+    {
+        if ( m_IsNotRestart )
+        {
+            // open file
+            std::ofstream ofs{m_path + m_rd_name + "." + m_extension, std::ofstream::out};
+            // write header row
+            ofs << "#";
+            ofs << "[1]step()";
+            ofs << m_sep;
+            ofs << "[2]time(s)";
+            ofs << m_sep;
+            ofs << "[3]" + reduction_type_string + " of " + parser_string + " (SI units)";
```

Comment:
`(SI units)` : we cannot write it, but this is true only if the expression for the diag makes sense (e.g., the user could ask for the max of `( (x + Ey)/sin(Bz^2) )` ;-) )

Reply:
True that the result can be in meaningless unit, but even in this case, `(SI units)` is a good reminder that the calculation is done with the fields and positions in SI units.

Reply:
Yes, I totally agree! It was a pointless observation...